spec/src/main/asciidoc/core/invokers.asciidoc

```diff
@@ -0,0 +1,173 @@
+Frameworks cannot invoke application methods directly, because they are not compiled against the application code.
+However, during application deployment, frameworks may observe application methods through CDI extensions and build an `Invoker` for each relevant method.
+The invokers can then be used at application runtime to invoke the methods indirectly.
+
+Method invokers are not supposed to be used by application code, as applications may invoke their own methods directly.
+
+[[building_invoker]]
+=== Building an `Invoker`
+
+The CDI container allows building an `Invoker` for a method of an <<enablement,enabled>> <<managed_beans,managed bean>>.
+The method for which the invoker is built is called the _target method_ of the invoker, and the managed bean is called the _target bean_ of the invoker.
+
+Invalid target methods are:
+
+* private methods,
+* constructors,
+* methods declared on the `java.lang.Object` class, except of the `toString()` method,
+* methods that are not declared on the bean class of the target bean or inherited from its supertypes.
+
+Attempting to build an invoker for an invalid target method leads to a deployment problem.
+
+Attempting to build an invoker for a non-static target method declared on a type that is not present in the set of bean types of the target bean leads to non-portable behavior.
+When the target bean is normal scoped, attempting to build an invoker for a non-static target method declared on an <<unproxyable,unproxyable>> bean type of the target bean leads to non-portable behavior.
+
+When the target bean is not a managed bean, attempting to build an invoker leads to a deployment problem.
+When the target bean is an interceptor, attempting to build an invoker leads to a deployment problem.
+
+Multiple managed beans may inherit a method from a common supertype.
+In that case, an invoker must be built for each target bean individually.
+An invoker built for one target bean may not be used to invoke the target method on an instance of another target bean.
+
+The only way to build an invoker is using the `InvokerBuilder`.
+An `InvokerBuilder` can only be obtained in CDI portable extensions and build compatible extensions.
+See <<invoker_builder>> for more information.
+
+[[using_invoker]]
+=== Using an `Invoker`
+
+The `Invoker` interface contains a single method:
+
+[source,java]
+----
+public interface Invoker<T, R> {
+    R invoke(T instance, Object[] arguments) throws Exception;
+}
+----
+
+Calling `invoke()` invokes the target method on given `instance` of the target bean, passing given `arguments`, and propagates back the return value or thrown exception.
+The `instance` and `arguments` may be contextual or non-contextual objects.
+
+Invoker implementations must be thread-safe.
+A single invoker instance may be used to perform multiple, possibly concurrent, invocations of the target method, possibly on different instances of the target bean, possibly with different arguments.
```

Comment:
I am not sure mentioning concurrent invocations is smart here.
While you can definitely do that, many beans are no built for concurrent usage by default (such as req. scoped beans) yet are a valid candidate for an invoker.  A user cannot know which beans will have invokers and so cannot guarantee concurrency works if attempted.
So, I'd either drop the note or add it into a separate sentence mentioning it's possible yet whether it's safe depends on the respective bean implementation.

Reply:
> I could add one more sentence if we wanted to be more clear about that:

:+1: sounds good

Reply:
Done.